File: src/main/java/io/lsdconsulting/lsd/distributed/captor/http/ResponseCaptor.java
package io.lsdconsulting.lsd.distributed.captor.http;

import feign.Response;
import io.lsdconsulting.lsd.distributed.captor.convert.TypeConverter;
import io.lsdconsulting.lsd.distributed.captor.header.HeaderRetriever;
import io.lsdconsulting.lsd.distributed.captor.http.derive.HttpStatusDeriver;
import io.lsdconsulting.lsd.distributed.captor.http.derive.PathDeriver;
import io.lsdconsulting.lsd.distributed.captor.http.derive.SourceTargetDeriver;
import io.lsdconsulting.lsd.distributed.captor.repository.InterceptedDocumentRepository;
import io.lsdconsulting.lsd.distributed.captor.repository.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.captor.repository.model.InterceptedInteractionFactory;
import io.lsdconsulting.lsd.distributed.captor.trace.TraceIdRetriever;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpResponse;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;

import static io.lsdconsulting.lsd.distributed.captor.repository.model.Type.RESPONSE;
import static org.apache.commons.lang3.StringUtils.EMPTY;

@RequiredArgsConstructor
@Slf4j
public class ResponseCaptor {

    private final InterceptedDocumentRepository interceptedDocumentRepository;
    private final InterceptedInteractionFactory interceptedInteractionFactory;
    private final SourceTargetDeriver sourceTargetDeriver;
    private final PathDeriver pathDeriver;
    private final TraceIdRetriever traceIdRetriever;
    private final HeaderRetriever headerRetriever;
    private final HttpStatusDeriver httpStatusDeriver;

    @SneakyThrows
    public InterceptedInteraction captureResponseInteraction(final Response response, Long elapsedTime) {
        final var requestHeaders = headerRetriever.retrieve(response.request());
        final var responseHeaders = headerRetriever.retrieve(response);
        final String path = pathDeriver.derivePathFrom(response.request().url());
        final String target = sourceTargetDeriver.deriveTarget(requestHeaders, path);
        final String serviceName = sourceTargetDeriver.deriveServiceName(requestHeaders);
        final String traceId = traceIdRetriever.getTraceId(requestHeaders);
        final String httpStatus = httpStatusDeriver.derive(response.status());
        final InterceptedInteraction interceptedInteraction = interceptedInteractionFactory.buildFrom(TypeConverter.convert(response.body()), requestHeaders, responseHeaders, traceId, serviceName, target, path, httpStatus, elapsedTime, null, RESPONSE);
        interceptedDocumentRepository.save(interceptedInteraction);
        return interceptedInteraction;
    }

    public InterceptedInteraction captureResponseInteraction(final HttpRequest request, final ClientHttpResponse response, final String target, final String path, final String traceId, Long elapsedTime) throws IOException {
        final var requestHeaders = headerRetriever.retrieve(request);
        final var responseHeaders = headerRetriever.retrieve(response);
        final String serviceName = sourceTargetDeriver.deriveServiceName(requestHeaders);
        final String body = copyBodyToString(response);
        final String httpStatus = response.getStatusCode().toString();
        final InterceptedInteraction interceptedInteraction = interceptedInteractionFactory.buildFrom(body, requestHeaders, responseHeaders, traceId, serviceName, target, path, httpStatus, elapsedTime, null, RESPONSE);
        interceptedDocumentRepository.save(interceptedInteraction);
        return interceptedInteraction;
    }

    private String copyBodyToString(final ClientHttpResponse response) throws IOException {
        if (response.getHeaders().getContentLength() == 0) {
            return EMPTY;
        }
        final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        final InputStream inputStream = response.getBody();
        inputStream.transferTo(outputStream);
        return outputStream.toString();
    }
}

File: spring-web/src/main/kotlin/io/lsdconsulting/lsd/distributed/interceptor/captor/ResponseCaptor.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor

import io.lsdconsulting.lsd.distributed.connector.model.InteractionType
import io.lsdconsulting.lsd.distributed.connector.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.SourceTargetDeriver
import io.lsdconsulting.lsd.distributed.interceptor.persistence.RepositoryService
import org.apache.commons.lang3.StringUtils
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpResponse
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.time.ZoneId
import java.time.ZonedDateTime

class ResponseCaptor(
    private val repositoryService: RepositoryService,
    private val sourceTargetDeriver: SourceTargetDeriver,
    private val httpHeaderRetriever: HttpHeaderRetriever,
    private val profile: String,
) {

    @Throws(IOException::class)
    fun captureResponseInteraction(
        request: HttpRequest,
        response: ClientHttpResponse,
        target: String,
        path: String,
        traceId: String,
        elapsedTime: Long
    ): InterceptedInteraction {
        val requestHeaders = httpHeaderRetriever.retrieve(request)
        val responseHeaders = httpHeaderRetriever.retrieve(response)
        val serviceName = sourceTargetDeriver.deriveServiceName(requestHeaders)
        val body = copyBodyToString(response)
        val httpStatus = response.statusCode.toString()
        val interceptedInteraction = buildInterceptedInteraction(
            target,
            path,
            traceId,
            elapsedTime,
            requestHeaders,
            responseHeaders,
            serviceName,
            body,
            httpStatus
        )
        repositoryService.enqueue(interceptedInteraction)
        return interceptedInteraction
    }

    private fun buildInterceptedInteraction(
        target: String,
        path: String,
        traceId: String,
        elapsedTime: Long,
        requestHeaders: Map<String, Collection<String>>,
        responseHeaders: Map<String, Collection<String>>,
        serviceName: String,
        body: String?,
        httpStatus: String?
    ) = InterceptedInteraction(
        traceId = traceId,
        body = body,
        requestHeaders = requestHeaders,
        responseHeaders = responseHeaders,
        serviceName = serviceName,
        target = target,
        path = path,
        httpStatus = httpStatus,
        httpMethod = null, interactionType = InteractionType.RESPONSE,
        profile = profile,
        elapsedTime = elapsedTime,
        createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
    )

    @Throws(IOException::class)
    private fun copyBodyToString(response: ClientHttpResponse): String {
        if (response.headers.contentLength == 0L) {
            return StringUtils.EMPTY
        }
        val outputStream = ByteArrayOutputStream()
        val inputStream = response.body
        inputStream.transferTo(outputStream)
        return outputStream.toString()
    }
}
